compiler/modules/front.syntax/token.bal

```diff
@@ -275,14 +285,50 @@ class Tokenizer {
         return err:syntax(msg, loc=err:location(self.file, self.currentPos()));
     }
 
+    function save() returns TokenizerState {
+        return {
+            lineIndex : self.lineIndex,
+            fragCodeIndex : self.fragCodeIndex,
+            codePointIndex : self.codePointIndex,
+            fragmentIndex : self.fragmentIndex,
+            tokenStartCodePointIndex : self.tokenStartCodePointIndex,
+            mode : self.mode,
+            curTok : self.curTok
+        };
+    }
+
+    function restore(TokenizerState s) {
+        self.lineIndex = s.lineIndex;
+        self.fragCodeIndex = s.fragCodeIndex;
+        self.codePointIndex = s.codePointIndex;
+        self.fragmentIndex = s.fragmentIndex;
+        self.tokenStartCodePointIndex = s.tokenStartCodePointIndex;
+        self.mode = s.mode;
+        self.curTok = s.curTok;
+
+        ScannedLine scannedLine = self.lines[self.lineIndex - 1];
+        self.fragCodes = scannedLine.fragCodes;
+        self.fragments = scannedLine.fragments;
+    }
+
 }
 
 function createPosition(int line, int column) returns Position {
     return (line << 32) | column;
 }
 
-public readonly class SourceFile {
+// JBUG when SourceFile is replaced with the ScannedSourceFile, get `incompatible types`
```

Comment:
This is the workaround, but I got some errors when I tried to use type inclusions. But it still works without that.